com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java

```diff
@@ -0,0 +1,133 @@
+import com.sap.cloud.lm.sl.common.ConflictException;
+
+@Named("applicationColorDetector")
+public class ApplicationColorDetector {
+
+    private static final ApplicationColor COLOR_OF_APPLICATIONS_WITHOUT_SUFFIX = ApplicationColor.BLUE;
+
+    @Inject
+    private OperationDao operationDao;
+
+    @Inject
+    private FlowableFacade flowableFacade;
+
+    public ApplicationColor detectLiveApplicationColor(DeployedMta deployedMta, String correlationId) {
+        if (deployedMta == null) {
+            return null;
+        }
+        ApplicationColor firstApplicationColor = getFirstApplicationColor(deployedMta);
+        Operation currentOperation = operationDao.find(correlationId);
+
+        Operation lastOperation = operationDao.find(createLatestOperationFilter(currentOperation.getMtaId(), currentOperation.getSpaceId()))
+            .get(0);
+
+        if (lastOperation.getState() != State.ABORTED) {
+            return firstApplicationColor;
+        }
+        String xs2BlueGreenDeployHistoricProcessInstanceId = flowableFacade.findHistoricProcessInstanceIdByProcessDefinitionKey(
+            lastOperation.getProcessId(), Constants.BLUE_GREEN_DEPLOY_SERVICE_ID);
+
+        ApplicationColor lastDeployedMtaColorValue = getColorFromHistoricProcess(xs2BlueGreenDeployHistoricProcessInstanceId);
+        Phase phase = getPhaseFromHistoricProcess(xs2BlueGreenDeployHistoricProcessInstanceId);
+
+        if (lastDeployedMtaColorValue == null || phase == null) {
+            return firstApplicationColor;
+        }
+        if (phase == Phase.UNDEPLOY) {
+            return lastDeployedMtaColorValue;
+        }
+        return firstApplicationColor;
+    }
+
+    public ApplicationColor detectSingularDeployedApplicationColor(DeployedMta deployedMta) {
+        if (deployedMta == null) {
+            return null;
+        }
+        ApplicationColor deployedApplicationColor = null;
+        for (DeployedMtaModule module : deployedMta.getModules()) {
+            ApplicationColor moduleApplicationColor = getApplicationColor(module);
+            if (deployedApplicationColor == null) {
+                deployedApplicationColor = (moduleApplicationColor);
+            }
+            if (deployedApplicationColor != moduleApplicationColor) {
+                throw new ConflictException(Messages.CONFLICTING_APP_COLORS, deployedMta.getMetadata()
+                    .getId());
+            }
+        }
+        return deployedApplicationColor;
+    }
+
+    private ApplicationColor getFirstApplicationColor(DeployedMta deployedMta) {
+        ApplicationColor firstApplicationColor = null;
+        Date firstApplicationColorCreatedOn = null;
+        for (DeployedMtaModule module : deployedMta.getModules()) {
+            Date moduleCreatedOn = module.getCreatedOn();
+            if (firstApplicationColorCreatedOn == null || firstApplicationColorCreatedOn.after(moduleCreatedOn)) {
+                firstApplicationColorCreatedOn = moduleCreatedOn;
```

Comment:
Same here with names - first should be earliest, Color can be dropped from name (e.g. 'earliestAppCreationDate'), switch 'firstApplicationColorCreatedOn.after(moduleCreatedOn)' check around (e.g. appCreatedOn.before(earliestAppCreationDate) or something similar

Reply:
Also could try rewriting the cycles as a stream/filter